File: Webserver/PythonLib/displaypage.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
r"""
  'displaypage' class receives a solr jsonRespone and 
  produces an html page, based on a template provided
  by an input file. 
   
"""
__author__ = 'bernardo gomez'
__date__ = 'august 2020'

import os
import sys
import json
import re
import biblio
import urllib.parse

class new_page:
   def __init__(self,html_file):
      """
         new_page receives an html_file and creates a corresponding 
         html_string.
      """
      self.html_string=""
      with open(html_file) as f:
            for line in f:
               self.html_string+=line
      return

   def __str__(self):
     """
        __str__ is a representation of html_string
     """
     return self.html_string

   def full_page(self,response,webserver_host,cgi_path,search_script):
     """
        full_page builds the html page that displays a single record.
        full_page receives 
          response: the solr description of the record,
          webserver_host: host name of the search platform,
          host_cgi: CGI path,
          search_script: CGI script
       full_page returns the html string and a return code.
                return code = 1 is failure,
                return code = 0 is success.
     """
     try:
          jsonResponse = response.json()
     except:
         print("connection to solr server failed ",file=sys.stderr)
         return "System error(31). Not a json response",1

     record_text=jsonResponse["response"]["docs"][0]["record"][0]
     try:
               online_access=jsonResponse["response"]["docs"][0]["link"][0]
     except:
                online_access=""
 # extract BIB elements from MARC record.
     marc=biblio.Record(record_text)
     title=marc.get_title()
     creator=marc.get_creator()
     contributor=marc.get_contributor()
     resource_type=marc.get_resource_type()
     publication_info=marc.get_publication_info()
     genre=marc.get_genre()
     topic=marc.get_topic()
     language=marc.get_language()
     oclc_link=marc.get_oclc_string()
     title=str(title)
     page_string=self.html_string
     page_string=page_string.replace("<!--TITLE=-->",title)
           #print_error(creator)

     if creator != "":
           user_query="\""+creator+"\""
           quote_user_query=urllib.parse.quote(user_query)
           search_url=webserver_host+cgi_path+search_script+"?query="+quote_user_query+"&fq=creator:"+quote_user_query
           main_author="<tr><th>Main Author:</th><td><span><a style=\"font-size: large\" href=\""+search_url+"\">"+creator+"</a></span></td></tr>" 
           page_string=page_string.replace("<!--MAIN_AUTHOR=-->",main_author)
     if len(contributor) > 0:
           contrib_list="<tr><th>Contributors:</th><td>"
           for this_contrib in contributor:
                   quote_user_query="\""+this_contrib+"\""
                   quote_user_query=urllib.parse.quote(quote_user_query)
                   search_url=webserver_host+cgi_path+search_script+"?query="+quote_user_query+"&fq=contributor:"+quote_user_query
                   item="<div><a style=\"font-size:large\" href=\""+search_url+"\">"+this_contrib+"</a></div>"
                   contrib_list+=item 
           contrib_list=contrib_list+"</td></tr>"
           page_string=page_string.replace("<!--OTHER_AUTHORS=-->",contrib_list)
     if len(genre)+ len(topic) > 0:
                subject_list=""
                subject_list="<tr><th>Subject:</th><td>"
                for g_item in genre:
                   query="\""+g_item+"\""
                   quote_query=urllib.parse.quote(query)
                   search_url=webserver_host+cgi_path+search_script+"?query="+quote_query+"&fq=genre:"+quote_user_query
                   item="<div><a style=\"font-size:large\" href=\""+search_url+"\">"+g_item+"</a></div>"
                   subject_list+=item 
                for t_item in topic:
                   query="\""+t_item+"\""
                   quote_query=urllib.parse.quote(query)
                   search_url=webserver_host+cgi_path+search_script+"?query="+quote_query+"&fq=topic:"+quote_user_query
                   item="<div><a style=\"font-size:large\" href=\""+search_url+"\">"+t_item+"</a></div>"
                   subject_list+=item 
                subject_list=subject_list+"</td></tr>"
                page_string=page_string.replace("<!--SUBJECTS=-->",subject_list)
     if online_access != "":
                online="<tr><th>Online access:</th><td>"
                item="<div><a style=\"font-size:large\" href=\""+online_access+"\" target=\"_blank\">Available</a></div>"
                online=online+item
                page_string=page_string.replace("<!--ONLINE_ACCESS=-->",online)
     if oclc_link != "":
                worldcat="<div><h2>"+"<a href=\""+oclc_link+"\" target=\"_blank\">"+"This item in Worldcat</a></h2></div>"
                page_string=page_string.replace("<!--WORLDCAT=-->",worldcat)
     staff_view=record_text
     page_string=page_string.replace("<!--STAFF_VIEW=-->",staff_view)
     try:
           authority_id=jsonResponse["response"]["docs"][0]["loc_authority_name"][0]
     except:
           authority_id=""
# http://worldcat.org/identities/lccn-n91005712/ <!--OCLC_IDENTITY_NETWORK=-->
     if authority_id != "":
               oclc_id_network="http://worldcat.org/identities/lccn-"+str(authority_id)+"/"
               id_network="<div><h2>"+"<a href=\""+oclc_id_network+"\" target=\"_blank\">"+"More about this author</a></h2></div>"
               page_string=page_string.replace("<!--OCLC_IDENTITY_NETWORK=-->",id_network)

     return page_string

   def dym_spell(self,response,user_query,webserver_host,cgi_path,search_script):
       """
        dym_spell builds the html page that displays a Did You Mean link.
        dym_spell receives 
          response: the solr description of the spelling suggestion,
          user_query: user query,
          webserver_host: host name of the search platform,
          host_cgi: CGI path,
          search_script: CGI script
       dym_spell returns the html string and a return code.
                return code = 1 is failure,
                return code = 0 is success.
       """
       page_string=self.html_string

       try:
          jsonResponse = response.json()
       except:
         print("connection to solr server failed ",file=sys.stderr)
         return "System error(21). Not a json response",1
  

       if jsonResponse["responseHeader"]["status"]  != 0:
             print ("solr error",file=sys.stderr)
             return "System error(6)",1

       in_line=re.compile(r'(.*<input type="text".*?value=")(.*?")(.*?>)(.*)',flags=re.DOTALL)

       if "spellcheck" not in jsonResponse:
              no_results='<div><h2>No Results Found</h2></div>'
              page_string=page_string.replace("<!--NORESULTS=-->",no_results)
              user_query=urllib.parse.unquote(user_query,encoding='utf-8')
              m=in_line.match(page_string)
              if m:
                 page_string=m.group(1)+user_query+'"'+m.group(3)+m.group(4)
              return page_string,0

       if len(jsonResponse["spellcheck"]["collations"]) == 0:
              no_results='<div><h2>No Results Found</h2></div>'
              page_string=page_string.replace("<!--NORESULTS=-->",no_results)
              user_query=urllib.parse.unquote(user_query,encoding='utf-8')
              m=in_line.match(page_string)
              if m:
                 page_string=m.group(1)+user_query+'"'+m.group(3)+m.group(4)
              return page_string,0
       suggested=jsonResponse["spellcheck"]["collations"][1]["collationQuery"]
       did_you_mean=str(suggested)
       search_uri=webserver_host+cgi_path+search_script+"?query="+did_you_mean
       did_you_mean='<div><span><h2>Did you mean <a href="'+search_uri+'">'+did_you_mean+'</a>?</h2></span></div>'
       user_query=urllib.parse.unquote(user_query,encoding='utf-8')
       m=in_line.match(page_string)
       if m:
                 page_string=m.group(1)+user_query+'"'+m.group(3)+m.group(4)
       page_string=page_string.replace("<!--DIDYOUMEAN=-->",did_you_mean)
       return page_string,0

   def record_list(self,response,user_query,webserver_host,cgi_path,search_script,refine_results,facet_script,page_size,record_count_marker,facet_marker,previous_marker,next_marker,full_record_script,results_marker):
      """
        record_list builds the html page that displays a list of records
            in brief format.
        record_list receives 
          response: the the solr description of records,
          user_query: user query,
          webserver_host: host name of the search platform,
          host_cgi: CGI path,
          search_script: CGI script that performs a solr search,
          refine_results: the html template for the facet display,
          facet_script: the script that searches a given facet,
          page_size: maximum number of records in a page,
          record_count_marker: place holder for number of records,
          facet_marker: place holder for facet display,
          previous_marker: place holder for link to previous page,
          next_marker: place holder for link to next page,
          full_record_script: CGI script that displays a single record,
          results_marker: place holder for record list,
        record_list returns the html string and a return code,
                return code = 1 is failure,
                return code = 0 is success.
      """
      try:
          jsonResponse = response.json()
      except:
         print("connection to solr server failed ",file=sys.stderr)
         return "System error(21). Not a json response",1

      refine_result_string=""
      number_found=jsonResponse["response"]["numFound"]
      start_record=jsonResponse["response"]["start"]

## read html template for facet list.
      with open(refine_results) as f:
            for line in f:
                refine_result_string+=line

      facet_info=jsonResponse["facet_counts"]
      facet_fields=facet_info["facet_fields"]
      facet_list={}
      for facet, value in facet_fields.items():
            state="get_item"
            item_list=[]
            string=""
            for item in value:
               if state == "get_item":
                  string=str(item)
                  state="get_count"
               else:
                  string=string+"_|_"+str(item)
                  item_list.append(string)
                  string=""
                  state="get_item"
            facet_list.update({facet:item_list})

      req_list=""
      for key,value in facet_list.items():
            html_marker="<!--"+str(key)+"=-->"
            for v in value:
               instance=v.split("_|_")
               if len(instance) == 2:
                  instance_value=instance[0]
                  instance_count=instance[1]
                  quote_user_query=urllib.parse.quote(user_query)
                  quote_instance_value=urllib.parse.quote(instance_value)
                  facet_query='&q='+quote_user_query+'&fq='+key+':'+str(quote_instance_value)
                  facet_request=webserver_host+cgi_path+facet_script+'?'+facet_query
                  req='<li><span><a href="'+facet_request+'">'+instance_value+'</a>'+'</span>'+'<span>'+"  "+str(instance_count)+'</span>'+'</li>'
                  req_list+=req
            refine_result_string=re.sub(html_marker,req_list,refine_result_string)
            req_list=""
      ###
##  solr found multiple matches. Display count number.
      result_count="Number of records: "+str(number_found)
      result_count="<h3>"+result_count+"</h3>"

##  solr found multiple matches. Display list of records.
      ## build results list as an HTML string
      result_page=self.html_string

      result_page=re.sub(record_count_marker,result_count,result_page)
      result_page=re.sub(facet_marker,refine_result_string,result_page)
#####
####
      record_table=""
      record_table='<div class="RecResults">'
      record_table+='<table class="RecordTable">'
      record_table+='<tbody>'
      for rec in jsonResponse["response"]["docs"]:
            record_id=rec["id"]
            record_text=rec["record"][0]
            marc=biblio.Record(record_text)
            #outcome,rec_title=get_title(record_text)
            rec_title=marc.get_title()
            #outcome,author_str=get_author(record_text)
            author_str=marc.get_creator()
            if author_str != "":
               author_str="by "+author_str
            publication_info=marc.get_publication_info()
            title_url=webserver_host+cgi_path+full_record_script+"?id="+str(record_id)
            record_title=rec_title
            record_table+='<tr>'
            record_table+='<td>'
            record_table+='<div class="SummaryFields">'
            record_table+='<h2>'
            record_table+='<a href="'+title_url+'">'+str(record_title)+'</a>'
            record_table+='</h2>'
            if author_str !=  "":
               record_table+='<h3>'+author_str+'</h3>'
            record_table+='<h3>'+publication_info+'</h3>'
            record_table+='</div>'
            record_table+='</td>'
            record_table+='</tr>'
      record_table+='</tbody>'
      record_table+='</table>'
      record_table+='</div>'
      result_page=re.sub(results_marker,record_table,result_page)

##  compute next_page and previous_page links
##  based on number_found, rows, start_record
##  if (start_record + rows) >=  number_found: no next_page
##  if (start_record -rows >= 0): create link to previous_page [start_record= start_record - rows]
##  if (start_record - rows) < 0 : no previous_page
##  if (start_record + rows) < number_found: create link to next_page [start_record= start_record+rows ]
##
##    result_page=re.sub(previous_marker,previouspage_element,result_page)
##    result_page=re.sub(next_marker,nextpage_element,result_page)
##
      nextpage_link=''
      previouspage_link=''
      previouspage_element=''
      nextpage_element=''
      if int(start_record) + int(page_size) < int(number_found):
        next_page=str(int(start_record)+int(page_size))
        quote_user_query=urllib.parse.quote(user_query)
        nextpage_link=webserver_host+cgi_path+'search?query='+quote_user_query+'&next='+next_page
        nextpage_element='<span>  <a href="'+nextpage_link+'" title="Next page"><b>NEXT</b></a>   </span>'
      if int(start_record) - int(page_size) >= 0:
        next_page=str(int(start_record)-int(page_size))
        previouspage_link=webserver_host+cgi_path+'search?query='+quote_user_query+'&next='+next_page
        previouspage_element='<span> <a href="'+previouspage_link+'" title="Previous page"><b>PREVIOUS</b></a>  </span>'

      result_page=re.sub(previous_marker,previouspage_element,result_page)
      result_page=re.sub(next_marker,nextpage_element,result_page)
      in_line=re.compile(r'(.*<input type="text".*?value=")(.*?")(.*?>)(.*)',flags=re.DOTALL)
      user_query=urllib.parse.unquote(user_query,encoding='utf-8')
      m=in_line.match(result_page)
      if m:
           result_page=m.group(1)+user_query+'"'+m.group(3)+m.group(4)

      return result_page,0

if __name__=="__main__":
    page=new_page("/home/bernardo/Webserver/Documents/euclid/full_record.html")
    print(page)
    

File: Webserver/Go-pkg/tinylib/displaypage/displaypage.go
// package displaypage builds an html string
package displaypage

import (
        "os"
        "bufio"
        "log"
  	"encoding/json"
  	"net/url"
	"errors"
	"strings"
	"regexp"
	"tinylib/biblio"
	"strconv"
        "io/ioutil"
)

type htmlPage  struct { 
    serverPage string    
}

// New receives a file with an html template and 
//   creates a corresponding html string.
func New(htmlFile string) htmlPage {
    var page htmlPage
    file,err:=os.Open(htmlFile)
    if err != nil {
        return  page
    }
    defer file.Close()
    page.serverPage=""
    scanner := bufio.NewScanner(file)
    for scanner.Scan() {
	page.serverPage+=scanner.Text()+"\n"
    }
    return page
}
// full_page builds the html page that displays a single record.
//      full_page receives 
//          jsonSolr: the solr description of the record,
//          webserverHost: host name of the search platform,
//          cgiPath: CGI path,
//          search_script: CGI script that performs generic search
//       full_page returns the html string and an error object.
//                error == nil, then success
//                error != nil, then failure.
func (page *htmlPage) FullPage(jsonSolr []byte, webserverHost string, cgiPath string, searchScript string) (string, error) {
     pageString:=""
     userQuery:=""
     l:=log.New(os.Stderr, "", 0)
     var jResponse map[string] interface{}
     err1:=json.Unmarshal(jsonSolr,&jResponse)
     if err1 != nil {
          return pageString,errors.New("Invalid json object")
     }
     jresponse,ok:=jResponse["response"].(map[string] interface{})
     if ! ok {
          return pageString,errors.New("Invalid json object")
     }
     jdocs,ok:=jresponse["docs"].([] interface{})
     if !ok {
          l.Println("solr didn't return docs")
          return pageString,errors.New("Invalid json object")
     }

     if len(jdocs) == 0 {
	    return pageString,errors.New("Marc record missing in json object.")
     }
     docList,ok:=jdocs[0].(map[string] interface {})
     if !ok {
          l.Println("solr didn't return docs")
          return pageString,errors.New("Invalid json object")
     }
     recordT,ok:=docList["record"].([] interface {})
     recordText:=recordT[0].(string)
     onlineAccess:=""
     authorityID:=""
     authorityName,ok:=docList["loc_authority_name"].([] interface {})
     if  ok {
	     authorityID=authorityName[0].(string)
     }
     marc:=biblio.New(recordText)
     title:=marc.GetTitle()
     creator:=marc.GetCreator()
     contributor:=marc.GetContributor()
     genre:=marc.GetGenre()
     topic:=marc.GetTopic()
     oclcString:=marc.GetOclcString()
     onlineAccess=marc.GetURL()
     pageString=page.serverPage
     pageString=strings.Replace(pageString,"<!--TITLE=-->",title,-1)
     if creator != "" {
	userQuery="\""+creator+"\""
	quoteUserQuery:=url.QueryEscape(userQuery)
	searchUrl:=webserverHost+cgiPath+searchScript+"?query="+quoteUserQuery+"&fq=creator:"+quoteUserQuery
	mainAuthor:="<tr><th>Main Author:</th><td><span><a href=\""+searchUrl+"\">"+creator+"</a></span></td></tr>" 
	pageString=strings.Replace(pageString,"<!--MAIN_AUTHOR=-->",mainAuthor,-1)
     }
     if len(contributor) > 0  {
           contribList:="<tr><th>Contributors:</th><td>"
	  for _,thisContrib:= range contributor {
	      quoteUserQuery:="\""+thisContrib+"\""
	      quoteUserQuery=url.QueryEscape(quoteUserQuery)
	      searchUrl:=webserverHost+cgiPath+searchScript+"?query="+quoteUserQuery+"&fq=contributor:"+quoteUserQuery
              item:="<div><a href=\""+searchUrl+"\">"+thisContrib+"</a></div>"
              contribList+=item
	  }
          contribList=contribList+"</td></tr>"
          pageString=strings.Replace(pageString,"<!--OTHER_AUTHORS=-->",contribList,-1)
     }

     if len(genre) + len(topic) > 0 {
	     query:=""
	     subjectList:="<tr><th>Subject:</th><td>"
	     for _,gItem := range genre {
		     query="\""+gItem+"\""
		     quoteQuery:=url.QueryEscape(query)
		     searchUrl:=webserverHost+cgiPath+searchScript+"?query="+quoteQuery+"&fq=genre:"+quoteQuery
		     item:="<div><a href=\""+searchUrl+"\">"+gItem+"</a></div>"
		     subjectList+=item
	     }
	     for _,tItem := range topic {
		     query="\""+tItem+"\""
		     quoteQuery:=url.QueryEscape(query)
		     searchUrl:=webserverHost+cgiPath+searchScript+"?query="+quoteQuery+"&fq=topic:"+quoteQuery
		     item:="<div><a href=\""+searchUrl+"\">"+tItem+"</a></div>"
		     subjectList+=item
	     }
	     subjectList=subjectList+"</td></tr>"
	     pageString=strings.Replace(pageString,"<!--SUBJECTS=-->",subjectList,-1)
     }
     if  onlineAccess != "" {
	     online:="<tr><th>Online access:</th><td>"
	     item:="<div><a href=\""+onlineAccess+"\">Available</a></div>"
	     online=online+item
	     pageString=strings.Replace(pageString,"<!--ONLINE_ACCESS=-->",online,-1)

     }
     if oclcString != "" {
	     worldCat:="<div><h2>"+"<a href=\""+oclcString+"\" target=\"_blank\">"+"This item in Worldcat</a></h2></div>"
	     pageString=strings.Replace(pageString,"<!--WORLDCAT=-->",worldCat,-1)
     }
     staffView:=recordText
     pageString=strings.Replace(pageString,"<!--STAFF_VIEW=-->",staffView,-1)
     if authorityID != "" {
	     oclcIdNetwork:="http://worldcat.org/identities/lccn-"+authorityID
	     idNetwork:="<div><h2>"+"<a href=\""+oclcIdNetwork+"\" target=_blank\">"+"More about this author</a></h2></div>"
	     pageString=strings.Replace(pageString,"<!--OCLC_IDENTITY_NETWORK=-->",idNetwork,-1)
     }
     return  pageString,nil
}


//      didYouMean builds the html page that displays a Did You Mean link.
//      didYouMean receives
//        jsonSlor: the solr description of the spelling suggestion,
//        userQuery: user query,
//        webserverHost: host name of the search platform,
//        cgiPath: CGI path,
//        search_script: CGI script that performs generic search
//        full_page returns the html string and an error object.
//                error == nil, then success
//                error != nil, then failure.
func (page *htmlPage) DidYouMean(jsonSolr [] byte, userQuery string,webserverHost string,cgiPath string,searchScript string) (string,error) {
	   pageString:=page.serverPage
	   l := log.New(os.Stderr, "", 0)
     var jResponse map[string] interface{}
     err1:=json.Unmarshal(jsonSolr,&jResponse)
     if err1 != nil {
          return pageString,errors.New("Invalid json object")
     }
	   header,ok:=jResponse["responseHeader"].(map[string] interface{})
           if !ok   {
//	        printError("Error. couldn't parse json response.")
		return "Error. couldn't parse json response.",errors.New("No Json response")
	   }
	   if header["status"].(float64) != 0 {
  	  	l.Println("status:",header["status"],)
//	        printError("Error. solr invalid request")
		return "Error. solr invalid request",errors.New("solr invalid request.")
	   }
	   jSpellCheck,ok:=jResponse["spellcheck"].(map[string] interface{})
	   if !ok {
 //        printError("FAILED jresponse!")
		return "FAILED jresponse",errors.New("FAILED response")
	   }
	   jCollations,ok:=jSpellCheck["collations"].([]  interface{})
	   if len(jCollations) == 0 {
		noResults:=`<div><h2>No Results Found</h2></div>`
	        pageString=strings.Replace(pageString,"<!--NORESULTS=-->",noResults,-1)
	        inLine,err:=regexp.Compile(`(.*<input type="text".*?value=")(.*?)(.*?)(.*)`)
	        if err != nil {
//	            printError("System error(35)")
		    return  "System error(35)",err
	        }
		pageString=replaceForm(inLine,pageString,userQuery)
		return pageString,nil
	   }
           collation:=jCollations[1].(map[string] interface{})
	   suggested:=collation["collationQuery"].(string)
	   searchURI:=webserverHost+cgiPath+searchScript+"?query="+suggested
           didYouMean:=`<div><span><h2>Did you mean <a href="`+searchURI+`">`+suggested+`</a>?</h2></span></div>`
	   pageString=strings.Replace(pageString,"<!--DIDYOUMEAN=-->",didYouMean,-1)
	   inLine,err:=regexp.Compile(`(.*<input type="text".*?value=")(.*?)(.*?)(.*)`)
	   if err != nil {
//	       l.Println("regexp compile failed",err)
//	       printError("System error(37)")
	       return "System error(37)",err
	   }
	   pageString=replaceForm(inLine,pageString,userQuery)
	return pageString,nil
}

// replaceForm places the user query into the search box of the html file.
// replaceForm receives:
// inPage: the regular expresion that matches the search box,
// pageString: the html page string,
// userQuery: the query string.
// replaceForm returns the  html page as a string.

func replaceForm(inPage *regexp.Regexp,pageString string,userQuery string) string{
      line:=strings.Split(pageString,"\n")
      page:=""
      for _,val:= range line {
          result:=inPage.FindStringSubmatch(val)
          if result != nil {
              page+=result[1]+result[2]+userQuery+result[4]+"\n"
           }  else {
              page+=val+"\n"
           }
      }
      return page
}

// MultipleRecs builds the html page that displays a list of records 
//  in brief format.
// MultipleRecs receives
//        jsonSolr: the the solr description of records,
//        userQuery: user query,
//        webserverHost: host name of the search platform,
//        cgiPath: CGI path,
//        searchScript: CGI script that performs a solr search,
//        refineResults: the html template for the facet display,
//        facetScript: the script that searches a given facet,
//        facetMarker: place holder for facet display,
//        recordCountMarker: place holder for number of records,
//        fullRecordScript: CGI script that displays a single record,
//        resultsMarker: place holder for record list,
//        pageSize: maximum number of records in a page,
//        previousMarker: place holder for link to previous page,
//        nextMarker: place holder for link to next page,
//  MultipleRecs returns the html string and an error object.
//                error == nil, then success
//                error != nil, then failure.

func (page *htmlPage) MultipleRecs(jsonSolr []byte, userQuery string, webserverHost string, cgiPath string, searchScript string, refineResults string, facetScript string, facetMarker string, recordCountMarker string, fullRecordScript string, resultsMarker string, pageSize string, previousMarker string, nextMarker string) (string, error) {
     pageString:=""
     l:=log.New(os.Stderr, "", 0)
     var jResponse map[string] interface{}
     err1:=json.Unmarshal(jsonSolr,&jResponse)
     if err1 != nil {
          return pageString,errors.New("Invalid json object")
     }
     jresponse,ok:=jResponse["response"].(map[string] interface{})
     if ! ok {
          return pageString,errors.New("Invalid json object")
     }
     numFound:=jresponse["numFound"].(float64)
     startRecord:=jresponse["start"].(float64)

     jdocs,ok:=jresponse["docs"].([] interface{})
     if !ok {
          l.Println("solr didn't return docs")
          return pageString,errors.New("Invalid json object")
     }


	// display multiple results
	refineResultString:=""
        data,err:=ioutil.ReadFile(refineResults)
	if err != nil {
	       l.Println("open failed",err)
	       return "System error(38)",err
        }
        refineResultString=string(data)
	facetInfo,ok:=jResponse["facet_counts"].(map[string] interface{})
	if !ok {
	       l.Println("json response doesn't have facet_counts")
	       return "System error(39)",errors.New("no facet counts in json")
	}
	facetFields,ok:=facetInfo["facet_fields"].(map[string] interface{})
	if !ok {
	       l.Println("facet_counts doesn't have facet_fields")
	       return "System error(40)",errors.New("facet_counts doesn't have facet_fields")
	}
// "facet_counts":{
//  "facet_queries":{},
//  "facet_fields":{
//    "pubDate":[
//      "1989",2,
//      "2004",2,
//      "1994",1,
//    "language":[
//      "english",5,
//      "chinese",1]},
//  "facet_ranges":{},
//  "facet_intervals":{},
//  "facet_heatmaps":{}}

	facets:=make(map[string] []string)
	var itemList[] string
	for fname,fList:= range facetFields {
           fvalue,ok:=fList.([] interface {})
	   //l.Println("facet name:",fname)
	   if !ok {
	       l.Println("error retrieving facet names")
	       return "System error(42)",errors.New("error retrieving facet names")
	   }
	   state:="get_item"
	   ivalue:=""
	   icount:=0.0
	   itemList=[]string{}
	   for _,v := range fvalue {
		if state == "get_item" {
		   ivalue=v.(string)
		   state="get_count"
		} else {
		   icount=v.(float64)
		   s64:=strconv.FormatFloat(icount, 'f', -1, 64)
		   is:=ivalue+"_|_"+s64
		   itemList=append(itemList,is)
		   state="get_item"
		}
	   }
	   //l.Println(instance)
           facets[fname]=itemList
	}
//      l.Println(facets)
        reqList:=""
	htmlMarker:=""
	var instance []string
	instanceValue:=""
	instanceCount:=""
	for key,ivalue:= range facets {
	    htmlMarker="<!--"+key+"=-->"
	    for _,iv:= range ivalue {
		   instance=strings.Split(iv,"_|_")
		   if len(instance) == 2{
		       instanceValue=instance[0]
		       instanceCount=instance[1]
	               quoteUserQuery:=url.QueryEscape(userQuery)
	               quoteInstanceValue:=url.QueryEscape(instanceValue)
  		       facetQuery:="&q="+quoteUserQuery+"&fq="+key+":"+quoteInstanceValue
//		       facetQuery:="&q="+quoteUserQuery+"&fq="+key+":"+instanceValue
		       facetRequest:=webserverHost+cgiPath+facetScript+"?"+facetQuery
		       // l.Println("facetQuery:",facetRequest)
		       spanE:=`<li><span><a href="`+facetRequest+`">`+instanceValue+`</a>`+`</span>`+`<span>`+" "+instanceCount+`</span>`+"</li>"
		       reqList+=spanE
		   }
            }
	    refineResultString=strings.Replace(refineResultString,htmlMarker,reqList,-1)
	    reqList=""
	}
	resultPage:=""
        resultPage=page.serverPage
	resultPage=strings.Replace(resultPage,facetMarker,refineResultString,-1)
	strNumFound:=strconv.FormatFloat(numFound, 'f', -1, 64)
        resultCount:=`Number of records:`+strNumFound
	resultCount=`<h3>`+resultCount+`</h3>`
	resultPage=strings.Replace(resultPage,recordCountMarker,resultCount,-1)
        recordTable:=""
	recordTable=`<div class="RecResults">`
	recordTable+=`<table class="RecordTable">`
	for _,docs:= range jdocs {
            rec:=docs.(map[string] interface{})
	    recordTxt:=rec["record"].([] interface{})
	    marcText:=recordTxt[0].(string)
	    marc:=biblio.New(marcText)
	    title:=marc.GetTitle()
	    //l.Println("title:",title)
	    authorStr:=marc.GetCreator()
	    if authorStr != "" {
		    authorStr="by "+authorStr
	    }
	    publicationInfo:=marc.GetPublicationInfo()
	    recordID:=marc.GetRecordID()
	    titleURL:=webserverHost+cgiPath+fullRecordScript+`?id=`+recordID
	    recordTable+=`<tr>`
	    recordTable+=`<td>`
            recordTable+=`<div class="SummaryFields">`
            recordTable+=`<h2>`
            recordTable+=`<a href="`+titleURL+`">`+title+`</a>`
            recordTable+=`</h2>`
	    if authorStr != ""{
		 recordTable+=`<h3>`+authorStr+`</h3>`
	    }
	    recordTable+=`<h3>`+publicationInfo+`</h3>`
	    recordTable+=`</div>`
            recordTable+=`</td>`
            recordTable+=`</tr>`
	}
	recordTable+=`</tbody>`
        recordTable+=`</table>`
        recordTable+=`</div>`
//	l.Println("almost there")
	resultPage=strings.Replace(resultPage,resultsMarker,recordTable,-1)
  //   	printPage(resultPage)
        nextPageLink:=""
        previousPageLink:=""
	nextPageElement:=""
	previousPageElement:=""
	startRecordInt:=int64(startRecord)
	pageSizeInt,err2:=strconv.Atoi(pageSize)
	if err2 != nil {
           return "System error(44)",errors.New("Page size conversion failed.")
	}
	numFoundInt:=int64(numFound)
//	if int(start_record) + int(page_size) < int(number_found):
        if startRecordInt + int64(pageSizeInt) <  numFoundInt {
	    nextPageInt:=startRecordInt + int64(pageSizeInt)
	    quotedQuery:=url.QueryEscape(userQuery)
	    nextPageStr:=strconv.FormatInt(nextPageInt,10)
	    nextPageLink=webserverHost+cgiPath+"search?query="+quotedQuery+"&next="+nextPageStr
	    nextPageElement=`<span>  <a href="`+nextPageLink+`" title="Next page"><b>NEXT</b></a>  </span>`
        }
	if startRecordInt - int64(pageSizeInt) >= 0 {
		 nextPageInt:=startRecordInt-int64(pageSizeInt)
	         nextPageStr:=strconv.FormatInt(nextPageInt,10)
	         quotedQuery:=url.QueryEscape(userQuery)
		 previousPageLink=webserverHost+cgiPath+"search?query="+quotedQuery+"&next="+nextPageStr
		 previousPageElement=`<span>  <a href="`+previousPageLink+`" title="Previous page"><b>PREVIOUS</b></a>  </span>`
        }
	resultPage=strings.Replace(resultPage,previousMarker,previousPageElement,-1)
	resultPage=strings.Replace(resultPage,nextMarker,nextPageElement,-1)
	inLine,err:=regexp.Compile(`(.*<input type="text".*?value=")(.*?)(.*?)(.*)`)
	if err != nil {
	         l.Println("regexp compile failed",err)
  	         return "System error(35)",errors.New("regexp compiled failed.")
	}
        resultPage=replaceForm(inLine,resultPage,userQuery)
        return resultPage,nil


     return pageString,nil
}
